File: ml-backend/day_night.py
"""Device-local day/night and calendar-day helpers.

Day is [DAY_START_HOUR, DAY_END_HOUR) in the *device* IANA timezone — never
UTC wall-clock and never the host or browser local zone. Desktop mirrors this
in dayNight.ts; cross-boundary tests assert identical verdicts.
"""

from __future__ import annotations

import os
from datetime import datetime, timezone
from zoneinfo import ZoneInfo, ZoneInfoNotFoundError

# Inclusive start, exclusive end — matches docs/dashboard.md and CROP_PROFILES.
DAY_START_HOUR = 6
DAY_END_HOUR = 18

DEFAULT_FALLBACK_TZ = "Africa/Johannesburg"


def default_device_timezone() -> str:
    """Configurable default for backfill / missing device.timezone.

    Prefer DIRT_SIGNAL_DEFAULT_TZ, else the host local zone name, else
    Africa/Johannesburg (UTC+2 deployment).
    """
    env = os.environ.get("DIRT_SIGNAL_DEFAULT_TZ", "").strip()
    if env:
        try:
            ZoneInfo(env)
            return env
        except ZoneInfoNotFoundError:
            pass
    try:
        local = datetime.now().astimezone().tzinfo
        if local is not None:
            key = getattr(local, "key", None)
            if isinstance(key, str) and key:
                ZoneInfo(key)
                return key
    except (ZoneInfoNotFoundError, Exception):
        pass
    return DEFAULT_FALLBACK_TZ


def ensure_aware_utc(value: datetime) -> datetime:
    if value.tzinfo is None:
        return value.replace(tzinfo=timezone.utc)
    return value.astimezone(timezone.utc)


def resolve_zone(tz_name: str | None) -> ZoneInfo:
    name = (tz_name or "").strip() or default_device_timezone()
    try:
        return ZoneInfo(name)
    except ZoneInfoNotFoundError:
        return ZoneInfo(default_device_timezone())


def local_hour(recorded_at: datetime, tz_name: str | None) -> int:
    """Hour 0–23 in the device timezone."""
    at = ensure_aware_utc(recorded_at)
    return at.astimezone(resolve_zone(tz_name)).hour


def local_day_key(recorded_at: datetime, tz_name: str | None) -> str:
    """YYYY-MM-DD in the device timezone."""
    at = ensure_aware_utc(recorded_at)
    local = at.astimezone(resolve_zone(tz_name))
    return local.strftime("%Y-%m-%d")


def is_day_period(recorded_at: datetime, tz_name: str | None) -> bool:
    """True when device-local hour is in [DAY_START_HOUR, DAY_END_HOUR)."""
    hour = local_hour(recorded_at, tz_name)
    return DAY_START_HOUR <= hour < DAY_END_HOUR


def is_night_period(recorded_at: datetime, tz_name: str | None) -> bool:
    """Complement of is_day_period (device-local)."""
    return not is_day_period(recorded_at, tz_name)

File: desktop/src/lib/dayNight.ts
/**
 * Device-local day/night and calendar-day helpers.
 *
 * Day is [DAY_START_HOUR, DAY_END_HOUR) in the *device* IANA timezone — never
 * UTC wall-clock and never the browser local zone. Backend mirrors this in
 * day_night.py; cross-boundary tests assert identical verdicts.
 */

/** Inclusive start, exclusive end — matches docs/dashboard.md. */
export const DAY_START_HOUR = 6;
export const DAY_END_HOUR = 18;

export const DEFAULT_DEVICE_TIMEZONE = "Africa/Johannesburg";

function toDate(recordedAt: string | Date): Date {
  return typeof recordedAt === "string" ? new Date(recordedAt) : recordedAt;
}

/**
 * Hour 0–23 in the device timezone via Intl (not Date#getHours).
 */
export function localHour(
  recordedAt: string | Date,
  timeZone: string,
): number {
  const d = toDate(recordedAt);
  const parts = new Intl.DateTimeFormat("en-US", {
    timeZone: timeZone || DEFAULT_DEVICE_TIMEZONE,
    hour: "numeric",
    hourCycle: "h23",
  }).formatToParts(d);
  const hourPart = parts.find((p) => p.type === "hour");
  const hour = hourPart ? Number.parseInt(hourPart.value, 10) : Number.NaN;
  if (Number.isNaN(hour)) {
    throw new Error(`Failed to resolve local hour for timezone ${timeZone}`);
  }
  // Some engines still emit "24" for midnight under h23; normalise.
  return hour === 24 ? 0 : hour;
}

/** YYYY-MM-DD in the device timezone. */
export function localDayKey(
  recordedAt: string | Date,
  timeZone: string,
): string {
  const d = toDate(recordedAt);
  const parts = new Intl.DateTimeFormat("en-CA", {
    timeZone: timeZone || DEFAULT_DEVICE_TIMEZONE,
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
  }).formatToParts(d);
  const year = parts.find((p) => p.type === "year")?.value;
  const month = parts.find((p) => p.type === "month")?.value;
  const day = parts.find((p) => p.type === "day")?.value;
  if (!year || !month || !day) {
    throw new Error(`Failed to resolve local day for timezone ${timeZone}`);
  }
  return `${year}-${month}-${day}`;
}

/** True when device-local hour is in [DAY_START_HOUR, DAY_END_HOUR). */
export function isDayPeriod(
  recordedAt: string | Date,
  timeZone: string,
): boolean {
  const hour = localHour(recordedAt, timeZone);
  return hour >= DAY_START_HOUR && hour < DAY_END_HOUR;
}

/** Complement of isDayPeriod (device-local). */
export function isNightPeriod(
  recordedAt: string | Date,
  timeZone: string,
): boolean {
  return !isDayPeriod(recordedAt, timeZone);
}
